editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,590 @@
+				}
+				draw_rect(Rect2(button_offset, 0, h, h), grabbed_col);
+				if (grabbed_col.r > 1 || grabbed_col.g > 1 || grabbed_col.b > 1) {
+					// Draw an indicator to denote that the currently selected color is "overbright".
+					draw_texture(get_theme_icon(SNAME("overbright_indicator"), SNAME("ColorPicker")), Point2(button_offset, 0));
+				}
 			} else {
-				draw_texture_rect(icon, Rect2(margin, margin, icon->get_width(), icon->get_height()));
+				// If no color is selected, draw grey color with 'X' on top.
+				draw_rect(Rect2(button_offset, 0, h, h), Color(0.5, 0.5, 0.5, 1));
+				draw_line(Vector2(button_offset, 0), Vector2(button_offset + h, h), Color(0.8, 0.8, 0.8));
+				draw_line(Vector2(button_offset, h), Vector2(button_offset + h, 0), Color(0.8, 0.8, 0.8));
+			}
+		} break;
+
+		case NOTIFICATION_VISIBILITY_CHANGED: {
+			if (!is_visible()) {
+				grabbing = GRAB_NONE;
 			}
 		} break;
 	}
 }
 
-Size2 GradientReverseButton::get_minimum_size() const {
-	return (get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons"))->get_size() + Size2(margin * 2, margin * 2));
+void GradientEditorRect::_on_mouse_exited() {
+	hovered_index = -1;
+	queue_redraw();
+}
+
+void GradientEditorRect::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_selected_index", "index"), &GradientEditorRect::set_selected_index);
+}
+
+GradientEditorRect::GradientEditorRect() {
+	set_focus_mode(FOCUS_ALL);
+
+	popup = memnew(PopupPanel);
+	picker = memnew(ColorPicker);
+	popup->add_child(picker);
+	popup->connect("about_to_popup", callable_mp(EditorNode::get_singleton(), &EditorNode::setup_color_picker).bind(GradientEditorRect::get_picker()));
+
+	preview_texture.instantiate();
+	preview_texture->set_width(1024);
+	add_child(popup, false, INTERNAL_MODE_FRONT);
+}
+
+///////////////////////
+
+void GradientEditor::_set_snap_enabled(bool p_enabled) {
+	gradient_editor_rect->set_snap_enabled(p_enabled);
+	snap_count_edit->set_visible(p_enabled);
+}
+
+void GradientEditor::_set_snap_count(int p_count) {
+	gradient_editor_rect->set_snap_count(CLAMP(p_count, 2, 100));
+}
+
+void GradientEditor::set_gradient(const Ref<Gradient> &p_gradient) {
+	gradient_editor_rect->set_gradient(p_gradient);
+}
+
+void GradientEditor::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE:
+		case NOTIFICATION_THEME_CHANGED: {
+			reverse_button->set_icon(get_theme_icon(SNAME("ReverseGradient"), SNAME("EditorIcons")));
+			snap_button->set_icon(get_theme_icon(SNAME("SnapGrid"), SNAME("EditorIcons")));
+		} break;
+		case NOTIFICATION_READY: {
+			Ref<Gradient> gradient = gradient_editor_rect->get_gradient();
+			// Set snapping settings based on the gradient.
+			if ((bool)gradient->get_meta(SNAME("_snap_enabled"), false) == true) {
+				snap_button->set_pressed(true);
+			}
+			if (gradient->has_meta(SNAME("_snap_count"))) {
+				snap_count_edit->set_value(gradient->get_meta(SNAME("_snap_count")));
+			}
+		} break;
```

Comment:
Adjusts the toggle button and the spinbox's value based on the gradient metadata.

I has to do this in READY, because otherwise the the `value_changed` signal on the spinbox wouldn't emit.